crypto/evp/p_lib.c

```diff
@@ -1459,3 +1459,131 @@ int evp_pkey_downgrade(EVP_PKEY *pk)
     return 0;     /* No downgrade, but at least the key is restored */
 }
 #endif  /* FIPS_MODE */
+
+const OSSL_PARAM *EVP_PKEY_gettable_params(EVP_PKEY *pkey)
+{
+    if (pkey == NULL
+        || pkey->keymgmt == NULL
+        || pkey->keydata == NULL)
+        return 0;
+    return evp_keymgmt_gettable_params(pkey->keymgmt);
+}
+
+/*
+ * This value is set into param->return_size to indicate if
+ * evp_keymgmt_get_params() ignored the parameter.
+ * The EVP_PKEY_get_XXXX_param() methods return 0 if the parameter was ignored.
+ */
+#define TERMINATOR ((size_t)-1)
+
+int EVP_PKEY_get_bn_param(EVP_PKEY *pkey, const char *key_name, BIGNUM **bn)
+{
+    unsigned char buffer[1024];
+    OSSL_PARAM params[2];
+
+    if (pkey == NULL
+        || pkey->keymgmt == NULL
+        || pkey->keydata == NULL
+        || key_name == NULL
+        || bn == NULL)
+        return 0;
+
+    memset(buffer, 0, sizeof(buffer));
+    params[0] = OSSL_PARAM_construct_BN(key_name, buffer, sizeof(buffer));
+    /* If the return_size is not changed from -1 then we know it was not found */
+    params[0].return_size = TERMINATOR;
```

Comment:
Although this is ugly it is the nicest way of doing it. Trying to do this on the destination is not practical.

Reply:
Hah, I'm doing something similar in another place, but I do `params[0].return_size = params[0].data_size + 1`, and later on determine if it's been touched with `!params[0].return_size != params[0].data_size + 1`.
The reason I did this is because theoretically, even though unlikely, `(size_t)-1` is a perfectly valid size, so if `params[0].data_size` happens to be that, you're screwed.
(with my formula, overflow isn't a concern)

Reply:
You could use `sizeof(buffer) + 1` instead of `params[0].data_size + 1` if you want a compile time constant.

Reply:
Ok done for all types..